Procedural Languages (PL/pgSQL + PL/Python) (#32)
IshaanAdarsh/Postgres-extension-tutorial: Quick-Start [DA]/plpgsql.md

# Procedural languages:

> Procedural languages in PostgreSQL provide a powerful way to extend the database's functionality by writing functions and procedures in languages other than SQL. These are natively supported by PostgreSQL and can be used directly within the database environment. These languages, also known as PL languages, offer developers the flexibility to implement complex business logic and computations directly within the database.

Reviewer: ```suggestion
Procedural languages in PostgreSQL provide a powerful way to extend the database's functionality by writing functions and procedures in languages other than SQL. These are natively supported by PostgreSQL and can be used directly within the database environment. These languages, also known as PLs, offer developers the flexibility to implement complex business logic and computations directly within the database.
```

Key points about procedural languages in PostgreSQL:

1. **Extended Functionality:** Procedural languages enable developers to write custom functions and procedures in languages like PL/pgSQL, PL/Python, PL/Tcl, PL/Perl, PL/R, and more. This allows for the creation of complex computations and data processing within the database.

2. **Control Flow and Logic:** Procedural languages offer control flow constructs like loops, conditionals, and exception handling, making it easier to implement business logic that goes beyond the capabilities of SQL.

> 3. **Custom Aggregates and Window Functions:** Procedural languages allow developers to define custom aggregates and window functions, enabling advanced data analysis and processing directly within the database.

Reviewer: Also triggers! Suggest linking to the docs for these features, in case the reader wonders what a custom aggregate or window function is and wants to find out.

4. **Performance Optimization:** By choosing the right procedural language for specific tasks, developers can achieve performance improvements over SQL, especially for complex data operations.

To use procedural languages in PostgreSQL:

1. **Enable the Language:** Before using a procedural language in a database, it needs to be enabled using the `CREATE LANGUAGE` command or by installing the relevant extensions.

2. **Write Functions and Procedures:** After enabling the language, developers can write functions and stored procedures using the chosen procedural language. This allows them to implement complex business logic, data transformations, and data analysis directly within the database.

## Quick Start: Creating a Postgres Extension using PL/pgSQL

In this quick start guide, we will create a simple PostgreSQL extension using [PL/pgSQL](https://www.postgresql.org/docs/current/plpgsql.html). The extension will include a PL/pgSQL function to subtract two numbers.

### Prerequisites

- [PostgreSQL](https://www.postgresql.org/download/) installed and running on your system.

> - Basic knowledge of SQL and PL/pgSQL programming.

Reviewer: Also need `make` and the Postgres libraries (dev package).

```sql
-- pg_plpgsql_ext--1.0.sql
-- Create the function that uses PL/pgSQL
CREATE OR REPLACE FUNCTION subtract_numbers(a integer, b integer)
RETURNS integer
LANGUAGE plpgsql
AS $$
BEGIN
    RETURN a - b;
END;
$$;
```

#### Explanation of the SQL Script:
- `CREATE OR REPLACE FUNCTION subtract_numbers(a integer, b integer) RETURNS integer LANGUAGE plpgsql AS $$`: This statement creates the function `subtract_numbers` with two integer arguments (`a` and `b`). The function will return an integer. The `LANGUAGE plpgsql` specifies that the function is written in PL/pgSQL.
- `BEGIN`: This is the beginning of the PL/pgSQL code block.
- `RETURN a - b;`: This line calculates the subtraction of `b` from `a` and returns the result.

### Step 3: Create the Control File:

Create a file named "pg_plpgsql_ext.control" in the "pg_plpgsql_ext" directory with the following content:

```control
# pg_plpgsql_ext.control

# Extension name
comment = 'A simple PostgreSQL extension using PL/pgSQL.'

# Default version of the extension

> default_version = '1.0'

Reviewer: I recommend using a semantic version at all times.

Author: Made all the changes.

# Extension is relocatable
relocatable = true
```

> - For more detailed information on Control Files, you can refer to the [PostgreSQL documentation on Control Files](https://www.postgresql.org/docs/current/extend-extensions.html#id-1.8.3.20.11).

Reviewer: You can also like to your other tutorial's section that introduces the control file.

> - For more detailed information on Control Files, you can refer to the [PostgreSQL documentation on Control Files](https://www.postgresql.org/docs/current/extend-extensions.html#id-1.8.3.20.11).

Author: ```suggestion
- For a basic explanation of Control file and its elements (LINK to QS)
- For more detailed information on Control Files, you can refer to the [PostgreSQL documentation on Control Files](https://www.postgresql.org/docs/current/extend-extensions.html#id-1.8.3.20.11).
```

#### Explanation of the Control File:
- `default_version = '1.0'`: This line specifies the default version of the extension.

> - `relocatable = true`: This line indicates that the extension is relocatable, allowing it to be moved to a different location.

Reviewer: ```suggestion
- `relocatable = true`: This line indicates that the extension is relocatable, meaning it does not have to live in a particular schema, and can be moved to a different schema if need be.
```

### Step 4: Create the Makefile:

Create a file named "Makefile" in the "pg_plpgsql_ext" directory with the following content:

```makefile
# Makefile for pg_plpgsql_ext
EXTENSION = pg_plpgsql_ext
DATA = pg_plpgsql_ext--1.0.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
```

> - For more information on the Makefiles, you can refer to the [PostgreSQL documentation on makefile](https://www.postgresql.org/docs/15/extend-pgxs.html).

Reviewer: Can also link to your earlier tutorial, yes?

> - For more information on the Makefiles, you can refer to the [PostgreSQL documentation on makefile](https://www.postgresql.org/docs/15/extend-pgxs.html).

Author: ```suggestion
- For a basic explanation of Makefile and its various components (LINK to QS)
- For more information on the Makefiles, you can refer to the [PostgreSQL documentation on makefile](https://www.postgresql.org/docs/15/extend-pgxs.html).
```

### Step 6: Enable the Extension in PostgreSQL

Connect to your PostgreSQL database using `psql`:

```bash
psql -U your_username -d your_database
```

Inside the PostgreSQL interactive terminal, enable the extension:

```sql
CREATE EXTENSION pg_plpgsql_ext;
```
- For more information on the `CREATE EXTENSION` command, you can refer to the [PostgreSQL documentation on `CREATE EXTENSION`](https://www.postgresql.org/docs/current/sql-createextension.html).

### Step 7: Test the Extension

Now that the extension is enabled, you can use the `subtract_numbers` function in SQL queries:

```sql
-- Example usage of the function
SELECT subtract_numbers(10, 5);

-- Output:
 subtract_numbers
-----------------
               5
(1 row)
```

## Quick Start: Creating a Postgres Extension using PL/Python:

In this quick start guide, we will create a simple PostgreSQL extension using [PL/Python](https://www.postgresql.org/docs/current/plpython.html). The extension will include a Python function to add two numbers.

### Prerequisites

- [PostgreSQL](https://www.postgresql.org/download/) installed and running on your system.

> - [plpython3u]((https://www.postgresql.org/docs/current/plpython.html)) procedural language available in your PostgreSQL installation.

Reviewer: Maybe mention what the `u` means.

> - [plpython3u]((https://www.postgresql.org/docs/current/plpython.html)) procedural language available in your PostgreSQL installation.

Author: ```suggestion
- [plpython3u]((https://www.postgresql.org/docs/current/plpython.html)) procedural language available in your PostgreSQL installation.

> In the context of PostgreSQL's procedural languages, the "u" in "plpython3u" stands for "untrusted". This signifies that the language implementation allows for executing arbitrary code and potentially accessing external resources.
```

> - Basic knowledge of [Python](https://www.python.org) programming.

Reviewer: Also `make` and the dev libraries.

> - Basic knowledge of [Python](https://www.python.org) programming.

Author: ```suggestion
- Basic knowledge of [Python](https://www.python.org) programming.
- `make` utility for building and compiling software. 
- Knowledge about the Postgres Development Libraries
```
